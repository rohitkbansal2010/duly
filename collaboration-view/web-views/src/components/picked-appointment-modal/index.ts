export * from './picked-appointment-modal';
