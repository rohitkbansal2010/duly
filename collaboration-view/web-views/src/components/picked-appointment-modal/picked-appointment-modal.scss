@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.customWidthModal {
  width: 44rem !important;
  max-width: 44rem !important;
}

.customBackDropModal {
  background-color: #0f172a !important;
  opacity: 0.15 !important;
}

.encounterModal {
  &Header {
    padding: 3.25rem 3.69rem 0.75rem !important;
    border-bottom: none !important;

    &Title {
      @include nexa-heavy-highlight;
      color: $dupage-magenta;
      font-size: 2.5rem !important;
    }
  }

  &Body {
    padding: 0 3.69rem !important;

    &Title {
      @include nexa-heavy-highlight;
      color: $dupage-violet;
      font-size: 1.5rem !important;
    }

    &List {
      padding: 2.31rem 0 0;
      list-style-type: none;

      &Item {
        display: flex;
        align-items: center;
        margin-bottom: 1.125rem;
        color: $dupage-violet;
        font-size: 1.25rem;
        font-family: "NexaRegular", sans-serif;
        line-height: 1.375rem;

        span {
          margin-left: 0.5rem;
        }

        & img {
          display: flex;
          align-items: center;
          justify-content: center;
          width: 1.69rem;
          height: 1.69rem;
        }
      }
    }
  }

  &Footer {
    padding: 1.69rem 3.69rem 3.25rem !important;
    border-top: none !important;

    &Cancel,
    &Launch {
      width: 10.75rem !important;
      height: 2.625rem !important;
      margin: 0 !important;
      color: red !important;
      font-weight: 800 !important;
      border-color: $dupage-violet !important;
      border-radius: 0.5rem !important;
      @include nexa-extra-bold-highlight;
    }

    &Cancel {
      margin-right: 1rem !important;
      color: $dupage-violet !important;
      background-color: $white !important;
    }

    &Launch {
      color: $white !important;
      background-color: $dupage-violet !important;
    }
  }
}
