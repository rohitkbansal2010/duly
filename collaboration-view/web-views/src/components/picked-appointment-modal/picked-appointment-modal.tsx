import { Button, Modal } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';

import { Avatar } from '@components/avatar';
import { PATIENT_GIVEN_NAMES_KEY, PRACTICIONER_KEY, WELCOME_SUFFIX_KEY } from '@constants';
import { CollaborationViewRoutes } from '@enums';
import { useAzureADAuth } from '@hooks';
import { clockVioletIcon, supervisedUserIcon } from '@icons';
import { setPickedAppointmentId } from '@redux/actions';
import { RootState } from '@redux/reducers';
import { AppDispatch } from '@redux/store';
import { AppointmentData } from '@types';
import {
  getUserRole,
  getSrcAvatar,
  showFirstPrefixItem,
  writeHumanName,
  writeTimeInterval,
  encodeURLParams,
  isTitleNotEmpty
} from '@utils';

import styles from './picked-appointment-modal.scss';

export const PickedAppointmentModal = () => {
  const dispatch: AppDispatch = useDispatch();

  const appointments = useSelector(({ TODAYS_APPOINTMENTS }: RootState) =>
    TODAYS_APPOINTMENTS.appointments);
  const pickedAppointmentId = useSelector(({ TODAYS_APPOINTMENTS }: RootState) =>
    TODAYS_APPOINTMENTS.pickedAppointmentId);

  const { logoutAzureADRedirect } = useAzureADAuth();

  // @TODO Fix typing on destructuring of probably undefined value
  const {
    patient: {
      patientGeneralInfo: {
        id: patientId,
        humanName: { familyName: patientFamilyName, givenNames },
      },
      isNewPatient,
    },
    timeSlot: { startTime, endTime },
    practitioner: {
      id: practitionerId,
      humanName: {
        familyName: practitionerFamilyName,
        givenNames: practitionerGivenNames,
        prefixes = [],
      },
      photo,
      role,
    },
    title,
  } = appointments.find(({ id }: AppointmentData) =>
    id === pickedAppointmentId);

  const launchAppointment = () => {
    sessionStorage.setItem(WELCOME_SUFFIX_KEY, isNewPatient ? 'to Duly' : 'Back');
    sessionStorage.setItem(PATIENT_GIVEN_NAMES_KEY, givenNames.join(' '));
    sessionStorage.setItem(PRACTICIONER_KEY, `${showFirstPrefixItem(prefixes)}${practitionerFamilyName}`);
    dispatch(setPickedAppointmentId(''));

    const [
      encodedAppointmentId,
      encodedPatientId,
      encodedPractitionerId,
    ] = encodeURLParams(pickedAppointmentId, patientId, practitionerId);

    logoutAzureADRedirect(`${CollaborationViewRoutes.welcome}/${encodedAppointmentId}/${encodedPatientId}/${encodedPractitionerId}`);
  };

  return (
    <Modal
      animation={false}
      show={true}
      onHide={() =>
        dispatch(setPickedAppointmentId(''))}
      className={styles.encounterModal}
      size="lg"
      centered
      dialogClassName={styles.customWidthModal}
      backdropClassName={styles.customBackDropModal}
    >
      <Modal.Header className={styles.encounterModalHeader}>
        <Modal.Title className={styles.encounterModalHeaderTitle}>
          {writeHumanName(givenNames, patientFamilyName)}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body className={styles.encounterModalBody}>
        <Modal.Title className={`${styles.encounterModalBodyTitle} text-nowrap`}>
          Are you sure you want to launch this appointment?
        </Modal.Title>
        <ul className={styles.encounterModalBodyList}>
          {practitionerFamilyName && (
            <li className={styles.encounterModalBodyListItem}>
              <Avatar
                src={getSrcAvatar(photo)}
                alt={`${showFirstPrefixItem(practitionerGivenNames)}${practitionerFamilyName}`.trim()}
                role={getUserRole({ role, prefixes })}
                width={1.69}
              />
              <span>{`${prefixes} ${practitionerFamilyName}`}</span>
            </li>
          )}
          {isTitleNotEmpty(title) && (
            <li className={styles.encounterModalBodyListItem}>
              <img src={supervisedUserIcon} alt="" />
              <span>{title}</span>
            </li>
          )}
          {startTime && endTime && (
            <li className={styles.encounterModalBodyListItem}>
              <img src={clockVioletIcon} alt="" />
              <span>{writeTimeInterval(new Date(startTime), new Date(endTime))}</span>
            </li>
          )}
        </ul>
      </Modal.Body>
      <Modal.Footer className={styles.encounterModalFooter}>
        <Button
          variant="secondary"
          onClick={() =>
            dispatch(setPickedAppointmentId(''))}
          className={styles.encounterModalFooterCancel}
        >
          Cancel
        </Button>
        <Button
          variant="primary"
          onClick={launchAppointment}
          className={styles.encounterModalFooterLaunch}
        >
          Launch
        </Button>
      </Modal.Footer>
    </Modal>
  );
};
