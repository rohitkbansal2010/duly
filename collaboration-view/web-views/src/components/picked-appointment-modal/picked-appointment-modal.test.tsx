import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ModalTitle, Button } from 'react-bootstrap';
import * as redux from 'react-redux';

import { appointmentsMockData } from '@mock-data';
import { writeHumanName, writeTimeInterval } from '@utils';

import { PickedAppointmentModal } from './picked-appointment-modal';

describe('PickedAppointmentModal', () => {
  let component: ReturnType<typeof shallow>;
  const mockPickedAppointment = appointmentsMockData[6];
  const {
    title: expectedServiceType,
    patient: {
      patientGeneralInfo: {
        humanName: {
          givenNames: patientGivenNames,
          familyName: patientFamilyName,
        },
      },
    },
    practitioner: {
      humanName: {
        familyName: providerFamilyName,
        prefixes: providerPrefixes,
      },
    },
    timeSlot: { startTime, endTime },
  } = mockPickedAppointment;
  beforeEach(() => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((selector: (state: unknown) => unknown) =>
        selector({
          TODAYS_APPOINTMENTS: {
            appointments: appointmentsMockData,
            pickedAppointmentId: mockPickedAppointment.id,
          },
        }));
    jest
      .spyOn(redux, 'useDispatch')
      .mockReturnValue(jest.fn());
    component = shallow(<PickedAppointmentModal />);
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  it('should exist', () => {
    expect(PickedAppointmentModal).toBeDefined();
  });
  it('should render PickedAppointmentModal', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
  it('should contains title', () => {
    const expectedTitle = 'Are you sure you want to launch this appointment?';
    expect(component.find(ModalTitle)).toHaveLength(2);
    expect(component.text().includes(expectedTitle)).toBeTruthy();
  });
  it('should contains patient name', () => {
    const expectedPatientName = writeHumanName(patientGivenNames || [], patientFamilyName);
    expect(component.text().includes(expectedPatientName)).toBeTruthy();
  });
  it('should contains service type of appointment', () => {
    expect(component.text().includes(expectedServiceType)).toBeTruthy();
  });
  it('should contains encounter start and end time', () => {
    const expectedTime = writeTimeInterval(new Date(startTime), new Date(endTime));
    expect(component.text().includes(expectedTime)).toBeTruthy();
  });
  it('should contains provider name', () => {
    const expectedProviderName = `${providerPrefixes} ${providerFamilyName}`;
    expect(component.text().includes(expectedProviderName)).toBeTruthy();
  });
  it('should contains cancel and launch buttons', () => {
    expect(component.find(Button)).toHaveLength(2);
    expect(component.find(Button).filterWhere(n =>
      n.text() === 'Cancel')).toHaveLength(1);
    expect(component.find(Button).filterWhere(n =>
      n.text() === 'Launch')).toHaveLength(1);
  });
});
