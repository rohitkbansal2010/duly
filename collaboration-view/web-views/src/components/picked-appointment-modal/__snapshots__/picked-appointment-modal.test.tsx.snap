// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickedAppointmentModal should render PickedAppointmentModal 1`] = `
<Modal
  animation={false}
  autoFocus={true}
  backdrop={true}
  backdropClassName="customBackDropModal"
  centered={true}
  className="encounterModal"
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    }
  }
  dialogClassName="customWidthModal"
  enforceFocus={true}
  keyboard={true}
  onHide={[Function]}
  restoreFocus={true}
  show={true}
  size="lg"
>
  <ModalHeader
    className="encounterModalHeader"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle
      className="encounterModalHeaderTitle"
    >
      Garcia Lucas
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    className="encounterModalBody"
  >
    <ModalTitle
      className="encounterModalBodyTitle text-nowrap"
    >
      Are you sure you want to launch this appointment?
    </ModalTitle>
    <ul
      className="encounterModalBodyList"
    >
      <li
        className="encounterModalBodyListItem"
      >
        <Avatar
          alt="Patricia Ling"
          role="MedicalDoctor"
          src=""
          width={1.69}
        />
        <span>
          Dr. Ling
        </span>
      </li>
      <li
        className="encounterModalBodyListItem"
      >
        <img
          alt=""
          src="supervised-user-violet.svg"
        />
        <span>
          Wellness Check
        </span>
      </li>
      <li
        className="encounterModalBodyListItem"
      >
        <img
          alt=""
          src="clock-violet.svg"
        />
        <span>
          10:30 am - 11:00 am
        </span>
      </li>
    </ul>
  </ModalBody>
  <ModalFooter
    className="encounterModalFooter"
  >
    <Button
      active={false}
      className="encounterModalFooterCancel"
      disabled={false}
      onClick={[Function]}
      variant="secondary"
    >
      Cancel
    </Button>
    <Button
      active={false}
      className="encounterModalFooterLaunch"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Launch
    </Button>
  </ModalFooter>
</Modal>
`;
